### PR TITLE
Adding the methods to ADO client for Batch Changes

### PR DESCRIPTION
**Pull request by varsanojidan**
Closes https://github.com/sourcegraph/sourcegraph/issues/46732
Don't panic, its a lot of tests and generated code :).

Mostly following the changes made in this [PR](https://github.com/sourcegraph/sourcegraph/pull/33944/files).

There are likely things(methods) that I will need that I missed due to my noobiness with BC, feel free to point them out, otherwise I'll find out soon enough when I attempt to implement the rest of it :P.

Adding methods to do ADO client that are required for Batch Changes

## Test plan

A lot of unit tests

### CODE REVIEW

File: internal/extsvc/azuredevops/testdata/golden/GetPullRequestStatuses.json [hunk 1]
```diff
@@ -0,0 +1 @@
+[]
```
Comment (varsanojidan) [inline internal/extsvc/azuredevops/testdata/golden/GetPullRequestStatuses.json:1]: Im still trying to setup a scenario for this one such that the response for the PR build statuses isn't empty, stay tuned.

File: internal/extsvc/azuredevops/pull_requests.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+	var pr PullRequestCommentResponse
+	if _, err = c.do(ctx, req, "", &pr); err != nil {
+		return PullRequestCommentResponse{}, err
+	}
+
+	return pr, nil
+}
+
+// CompletePullRequest abandons (closes) the specified PR, returns the updated PR. The PullRequestUpdateInput input just needs to specify the LastMergeSourceCommit.ID.
```
Comment (eseliger) [inline internal/extsvc/azuredevops/pull_requests.go:158]: abandon? not complete?

**Review by eseliger — APPROVED** [review]
looks solid, let's go! we'll find the quirks with their API once we try to wire all the things up 😆

File: internal/extsvc/azuredevops/pull_requests.go (hunk 2 above)
Comment (varsanojidan) [reply internal/extsvc/azuredevops/pull_requests.go]: shhhhhhhh you saw nothing 👀

File: internal/extsvc/azuredevops/pull_requests.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	}
+
+	return pr, nil
+}
+
+// CreatePullRequest creates a new PR with the specified properties, returns the newly created PR.
+// NOTE: this API needs repository ID specified not repository Name in OrgProjectRepoArgs.
+func (c *Client) CreatePullRequest(ctx context.Context, args OrgProjectRepoArgs, input CreatePullRequestInput) (PullRequest, error) {
+	queryParams := make(url.Values)
+	queryParams.Set("api-version", apiVersion)
```
Comment (mrnugget) [inline internal/extsvc/azuredevops/pull_requests.go:44]: Nit, but couldn't we set the api version in `do()`?

File: internal/extsvc/azuredevops/repositories.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	return repo, nil
+}
+
+func (c *Client) GetCommitForRepositoryBranchHead(ctx context.Context, args OrgProjectRepoArgs, branchName string) (Ref, error) {
+	var allRefs []Ref
+	continuationToken := ""
+
+	for {
+		queryParams := make(url.Values)
+		queryParams.Set("api-version", apiVersion)
```
Comment (mrnugget) [inline internal/extsvc/azuredevops/repositories.go:82]: Nit: do you need to re-recreate the `queryParams` in every iteration or can you reuse them and only update the `continuationToken`?

File: internal/extsvc/azuredevops/repositories.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+		// The filter here by branch name is only a substring match, so we aren't guaranteed to only get one result.
+		queryParams.Set("filter", fmt.Sprintf("heads/%s", branchName))
+		if continuationToken != "" {
+			queryParams.Set("continuationToken", continuationToken)
+		}
+		urlRepositoriesByProjects := url.URL{Path: fmt.Sprintf("%s/%s/_apis/git/repositories/%s/refs", args.Org, args.Project, args.RepoNameOrID), RawQuery: queryParams.Encode()}
```
Comment (mrnugget) [inline internal/extsvc/azuredevops/repositories.go:88]: I'd pull the `Path` out into a local var, out of the loop. Goal would be to keep the loop body as small as possible, so it's easy to see what's going ("we use the same request, but change the continuation token, to paginate")

File: internal/extsvc/azuredevops/repositories.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+		}
+	}
+
+	for _, ref := range allRefs {
+		if ref.Name == fmt.Sprintf("refs/heads/%s", branchName) {
+			return ref, nil
+		}
+	}
+
+	return Ref{}, errors.New("branch not found")
```
Comment (mrnugget) [inline internal/extsvc/azuredevops/repositories.go:113]:
```suggestion
	return Ref{}, errors.Newf("branch %q not found", branchName)
```

**Review by mrnugget — APPROVED** [review]
Code machine!

**Review by indradhanush — APPROVED** [review]
Nit: I think it might be simpler to have all the client methods in a single `client.go` file to make it easier to not have to remember which file I want to open while exploring this code based on the client method's name.

**Comment by eseliger** [comment]
> Nit: I think it might be simpler to have all the client methods in a single client.go file to make it easier to not have to remember which file I want to open while exploring this code based on the client method's name.

personal opinion: I find this actually much more readable because it doesn’t mix like 25 concerns in a single file 😅 I’ll use search or go to der anyways to end up here, so that is not a big concern to me. But everyone uses their IDE differently and I don’t feel strongly about this 🙂

**Comment by varsanojidan** [comment]
@indradhanush It actually used to be that all the methods were in `client.go` but tbh I was inspired by the linked PR above I thought it was quite a bit neater to have the related methods in separate files to match what they are modifying, so personally I agree with Erik here. Do you feel strongly that it should be all in `client.go`?

**Pull request merged**